lib/Actions.js

```diff
@@ -76,34 +99,68 @@ Actions.prototype = {
         debug('Perform action %o for generated selector %s', action, selector);
 
         var waitingForPage;
-        if (action.waitForPage) {
+        if (action.waitForPage || action.type === this.TYPES.BACK) {
             waitingForPage = this.waitForPage(action.waitForPageTimeout);
         } else {
             waitingForPage = vow.resolve();
         }
 
+        var casesPromise;
+        if (action.cases) {
+            casesPromise = this._performCases(action.cases, parentSelector);
+        }
+
         var actionPromise;
         switch (action.type) {
             case this.TYPES.CLICK:
                 actionPromise = this.click(selector);
                 break;
 
+            case this.TYPES.MOUSE_DOWN:
+                actionPromise = this.mousedown(selector);
+                break;
+
+            case this.TYPES.MOUSE_UP:
+                actionPromise = this.mouseup(selector);
+                break;
+
             case this.TYPES.WAIT:
                 actionPromise = this.waitElement(selector, action.timeout);
                 break;
 
+            case this.TYPES.WAIT_FOR_VISIBLE:
+                actionPromise = this.waitElementIsVisible(selector, action.timeout);
+                break;
+
+            case this.TYPES.WAIT_FOR_PATTERN:
+                actionPromise = this.waitForPattern(selector, action.pattern, action.timeout);
+                break;
+
+            case this.TYPES.WAIT_FOR_PAGE:
+                actionPromise = this.waitForPage(action.timeout);
+                break;
+
             case this.TYPES.TYPE:
                 actionPromise = this.type(selector, action.text);
                 break;
 
             case this.TYPES.CONDITION:
-                actionPromise = this.performConditionalActions(selector, action.conditions, action.actions);
+                actionPromise = this.performConditionalActions(selector, action.conditions, action.actions, action.elseActions);
                 break;
 
             case this.TYPES.EXIST:
                 actionPromise = this.exist(selector);
                 break;
 
+            case this.TYPES.BACK:
+                actionPromise = this.back();
+                break;
+
+            case this.TYPES.PROVIDE_COLLECTION:
+                debug('Providing collection %o', action.collection);
+                actionPromise = vow.resolve(action.collection);
```

Comment:
I guess, better give an ability to return not only collection, but full `{Rule}`. That will be possible to use that kind of action for getting any rule type with scope, or use parent scope.